File: ohara-kafka/src/main/scala/com/island/ohara/kafka/KafkaClient.scala
package com.island.ohara.kafka

import java.util
import java.util.Properties
import java.util.concurrent.TimeUnit

import com.island.ohara.client.util.CloseOnce
import com.island.ohara.kafka.KafkaClient._
import org.apache.kafka.clients.CommonClientConfigs
import org.apache.kafka.clients.admin.{AdminClient, NewPartitions, NewTopic}
import org.apache.kafka.common.KafkaFuture
import org.apache.kafka.common.config.ConfigResource
import org.apache.kafka.common.errors.UnknownTopicOrPartitionException

import scala.concurrent.ExecutionException
import scala.concurrent.duration._

/**
  * a helper methods used by configurator. It provide many helper method to operate kafka cluster.
  */
trait KafkaClient extends CloseOnce {

  def topicCreator(): TopicCreator

  def exist(topicName: String, timeout: Duration = DEFAULT_TIMEOUT): Boolean

  def nonExist(topicName: String, timeout: Duration = DEFAULT_TIMEOUT): Boolean = !exist(topicName, timeout)

  /**
    * describe the topic existing in kafka. If the topic doesn't exist, exception will be thrown
    * @param topicName topic name
    * @param timeout timeout
    * @return TopicDescription
    */
  def topicDescription(topicName: String, timeout: Duration = DEFAULT_TIMEOUT): TopicDescription

  def addPartitions(topicName: String, numberOfPartitions: Int, timeout: Duration = DEFAULT_TIMEOUT): Unit

  def deleteTopic(topicName: String, timeout: Duration = DEFAULT_TIMEOUT): Unit

  def listTopics(timeout: Duration = DEFAULT_TIMEOUT): Seq[String]

  def brokers: String

  def consumerBuilder(): ConsumerBuilder
}

object KafkaClient {
  private val DEFAULT_TIMEOUT = 10 seconds

  /**
    * this impl will host a kafka.AdminClient so you must call the #close() to release the kafka.AdminClient.
    *
    * @param _brokers the kafka brokers information
    * @return a impl from KafkaClient
    */
  def apply(_brokers: String): KafkaClient = new KafkaClient() {
    private[this] val admin = AdminClient.create(toAdminProps(_brokers))

    override def exist(topicName: String, timeout: Duration): Boolean =
      admin
        .listTopics()
        .names()
        .thenApply(new KafkaFuture.Function[util.Set[String], Boolean] {
          override def apply(a: util.Set[String]): Boolean = a.contains(topicName)
        })
        .get(timeout.toMillis, TimeUnit.MILLISECONDS)

    override protected def doClose(): Unit = admin.close()

    import scala.collection.JavaConverters._
    override def topicCreator(): TopicCreator = new TopicCreator {
      override protected def doCreate(request: TopicCreator.Request): Unit = admin
        .createTopics(
          util.Arrays.asList(new NewTopic(request.name, request.numberOfPartitions, request.numberOfReplications)
            .configs(request.options.asJava)))
        .values()
        .get(request.name)
        .get(request.timeout.toMillis, TimeUnit.MILLISECONDS)
    }

    override def topicDescription(topicName: String, timeout: Duration): TopicDescription =
      try {
        val configKey = new ConfigResource(ConfigResource.Type.TOPIC, topicName)
        val options = admin
          .describeConfigs(util.Arrays.asList(configKey))
          .values()
          .get(configKey)
          .get()
          .entries()
          .asScala
          .map(
            o =>
              TopicOption(key = o.name(),
                          value = o.value(),
                          isDefault = o.isDefault,
                          isSensitive = o.isSensitive,
                          isReadOnly = o.isReadOnly))
          .toSeq
        Option(admin.describeTopics(util.Arrays.asList(topicName)).values().get(topicName))
          .map(_.get(timeout.toMillis, TimeUnit.MILLISECONDS))
          .map(topicPartitionInfo =>
            TopicDescription(
              topicPartitionInfo.name(),
              topicPartitionInfo.partitions().size(),
              // TODO: seems it has chance that each partition has different number from replications. by chia
              topicPartitionInfo.partitions().get(0).replicas().size().toShort,
              options
          ))
          .getOrElse(throw new IllegalArgumentException(s"the topic:$topicName isn't existed"))
      } catch {
        case e: ExecutionException =>
          e.getCause match {
            // substitute None for UnknownTopicOrPartitionException
            case _: UnknownTopicOrPartitionException =>
              throw new IllegalArgumentException(s"the $topicName doesn't exist")
            case other: Throwable => throw other
          }
      }

    override def addPartitions(topicName: String, numberOfPartitions: Int, timeout: Duration): Unit = {
      val current = topicDescription(topicName, timeout)
      if (current.numberOfPartitions > numberOfPartitions)
        throw new IllegalArgumentException("Reducing the number from partitions is disallowed")
      if (current.numberOfPartitions < numberOfPartitions) {
        import scala.collection.JavaConverters._
        admin
          .createPartitions(Map(topicName -> NewPartitions.increaseTo(numberOfPartitions)).asJava)
          .all()
          .get(timeout.toMillis, TimeUnit.MILLISECONDS)
      }
    }

    override def deleteTopic(topicName: String, timeout: Duration): Unit =
      admin.deleteTopics(util.Arrays.asList(topicName)).all().get(timeout.toMillis, TimeUnit.MILLISECONDS)

    override def listTopics(timeout: Duration): Seq[String] =
      admin.listTopics().names().get(timeout.toMillis, TimeUnit.MILLISECONDS).asScala.toList
    override def brokers: String = _brokers
    override def consumerBuilder(): ConsumerBuilder = new ConsumerBuilder().brokers(brokers)

  }

  /**
    * a helper method used to put the brokers information to java properties.
    * Usually controlling the admin client only require the broker information.
    *
    * @param brokers kafka brokers information
    * @return a properties with brokers information
    */
  private[this] def toAdminProps(brokers: String): Properties = {
    val adminProps = new Properties()
    adminProps.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, brokers)
    adminProps
  }
}
case class TopicOption(key: String, value: String, isDefault: Boolean, isSensitive: Boolean, isReadOnly: Boolean)
case class TopicDescription(name: String,
                            numberOfPartitions: Int,
                            numberOfReplications: Short,
                            options: Seq[TopicOption])

File: ohara-kafka/src/main/java/com/island/ohara/kafka/KafkaClient.java
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.kafka;

import com.island.ohara.common.util.Releasable;
import com.island.ohara.kafka.exception.*;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.stream.Collectors;
import org.apache.kafka.clients.CommonClientConfigs;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.NewPartitions;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.common.KafkaFuture;
import org.apache.kafka.common.config.ConfigResource;
import org.apache.kafka.common.errors.UnknownTopicOrPartitionException;

/**
 * a helper methods used by configurator. It provide many helper method to operate kafka cluster.
 *
 * <p>Every method with remote call need to overload in Default Timeout
 */
public interface KafkaClient extends Releasable {
  Duration DEFAULT_TIMEOUT = Duration.ofSeconds(10);

  TopicCreator topicCreator();

  boolean exist(String topicName, Duration timeout);

  default boolean exist(String topicName) {
    return exist(topicName, DEFAULT_TIMEOUT);
  }

  default boolean nonExist(String topicName, Duration timeout) {
    return !exist(topicName, timeout);
  }

  default boolean nonExist(String topicName) {
    return !exist(topicName, DEFAULT_TIMEOUT);
  }

  /**
   * describe the topic existing in kafka. If the topic doesn't exist, exception will be thrown
   *
   * @param topicName topic name
   * @param timeout timeout
   * @return TopicDescription
   */
  TopicDescription topicDescription(String topicName, Duration timeout);

  default TopicDescription topicDescription(String topicName) {
    return topicDescription(topicName, DEFAULT_TIMEOUT);
  }

  void addPartitions(String topicName, int numberOfPartitions, Duration timeout);

  default void addPartitions(String topicName, int numberOfPartitions) {
    addPartitions(topicName, numberOfPartitions, DEFAULT_TIMEOUT);
  }

  void deleteTopic(String topicName, Duration timeout);

  default void deleteTopic(String topicName) {
    deleteTopic(topicName, DEFAULT_TIMEOUT);
  }

  List<String> listTopics(Duration timeout);

  default List<String> listTopics() {
    return listTopics(DEFAULT_TIMEOUT);
  }

  String brokers();

  ConsumerBuilder consumerBuilder();

  static KafkaClient of(String outerbrokers) {
    return new KafkaClient() {

      private final String brokers = outerbrokers;

      private final AdminClient admin = AdminClient.create(toAdminProps(outerbrokers));

      private final ExceptionHandler handler =
          ExceptionHandler.creator()
              .add(ExecutionException.class, (e) -> new OharaExecutionException(e.getCause()))
              .add(InterruptedException.class, OharaInterruptedException::new)
              .add(TimeoutException.class, OharaTimeoutException::new)
              .create();

      @Override
      public TopicCreator topicCreator() {
        return new TopicCreator() {

          @Override
          public void create(String name) {
            CheckedExceptionUtil.wrap(
                () ->
                    admin
                        .createTopics(
                            Collections.singletonList(
                                new NewTopic(name, numberOfPartitions, numberOfReplications)
                                    .configs(options)))
                        .values()
                        .get(name)
                        .get(timeout.toMillis(), TimeUnit.MILLISECONDS),
                handler);
          }
        };
      }

      @Override
      public boolean exist(String topicName, Duration timeout) {

        return CheckedExceptionUtil.wrap(
            () ->
                admin
                    .listTopics()
                    .names()
                    .thenApply(
                        new KafkaFuture.Function<Set<String>, Boolean>() {
                          @Override
                          public Boolean apply(Set<String> strings) {
                            return strings.contains(topicName);
                          }
                        })
                    .get(timeout.toMillis(), TimeUnit.MILLISECONDS),
            handler);
      }

      @Override
      public TopicDescription topicDescription(String topicName, Duration timeout) {
        return CheckedExceptionUtil.wrap(
            () -> {
              try {
                ConfigResource configKey = new ConfigResource(ConfigResource.Type.TOPIC, topicName);
                List<TopicOption> options =
                    admin
                        .describeConfigs(Collections.singletonList(configKey))
                        .values()
                        .get(configKey)
                        .get()
                        .entries()
                        .stream()
                        .map(
                            o ->
                                new TopicOption(
                                    o.name(),
                                    o.value(),
                                    o.isDefault(),
                                    o.isSensitive(),
                                    o.isReadOnly()))
                        .collect(Collectors.toList());

                org.apache.kafka.clients.admin.TopicDescription topicPartitionInfo =
                    Optional.ofNullable(
                            admin
                                .describeTopics(Collections.singletonList(topicName))
                                .values()
                                .get(topicName))
                        .orElseThrow(
                            () ->
                                new IllegalArgumentException(
                                    String.format("the topic:%s isn't existed", topicName)))
                        .get(timeout.toMillis(), TimeUnit.MILLISECONDS);

                return new TopicDescription(
                    topicPartitionInfo.name(),
                    topicPartitionInfo.partitions().size(),
                    (short) topicPartitionInfo.partitions().get(0).replicas().size(),
                    options);

              } catch (ExecutionException e) {
                if (e.getCause() instanceof UnknownTopicOrPartitionException)
                  throw new IllegalArgumentException(
                      String.format("the %s doesn't exist", topicName));
                throw e;
              }
            },
            handler);
      }

      @Override
      public void addPartitions(String topicName, int numberOfPartitions, Duration timeout) {
        TopicDescription current = topicDescription(topicName, timeout);
        if (current.numberOfPartitions() > numberOfPartitions)
          throw new IllegalArgumentException("Reducing the number from partitions is disallowed");
        if (current.numberOfPartitions() < numberOfPartitions) {
          Map<String, NewPartitions> map = new HashMap<>();
          map.put(topicName, NewPartitions.increaseTo(numberOfPartitions));

          CheckedExceptionUtil.wrap(
              () ->
                  admin.createPartitions(map).all().get(timeout.toMillis(), TimeUnit.MILLISECONDS),
              handler);
        }
      }

      @Override
      public void deleteTopic(String topicName, Duration timeout) {
        CheckedExceptionUtil.wrap(
            () ->
                admin
                    .deleteTopics(Collections.singletonList(topicName))
                    .all()
                    .get(timeout.toMillis(), TimeUnit.MILLISECONDS),
            handler);
      }

      @Override
      public List<String> listTopics(Duration timeout) {
        return CheckedExceptionUtil.wrap(
            () ->
                new ArrayList<>(
                    admin.listTopics().names().get(timeout.toMillis(), TimeUnit.MILLISECONDS)),
            handler);
      }

      @Override
      public String brokers() {
        return brokers;
      }

      @Override
      public ConsumerBuilder consumerBuilder() {
        return new ConsumerBuilder().brokers(brokers);
      }

      /**
       * this impl will host a kafka.AdminClient so you must call the #close() to release the
       * kafka.AdminClient.
       *
       * @param brokers the kafka brokers information
       * @return a impl from KafkaClient
       */
      private Properties toAdminProps(String brokers) {
        Properties adminProps = new Properties();
        adminProps.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, brokers);
        return adminProps;
      }

      @Override
      public void close() {
        admin.close();
      }
    };
  }
}
